src/Agent/NewRelic/Agent/Core/Agent.cs

```diff
@@ -437,15 +437,21 @@ public void RecordLogMessage(string frameworkName, object logEvent, Func<object,
                 if (transaction != null && transaction.IsValid)
                 {
                     // use transaction batching for messages in transactions
-                    transaction.LogEvents.Add(new LogEventWireModel(timestamp, logMessage, normalizedLevel, spanId, traceId));
+                    if (!transaction.AddLogEvent(new LogEventWireModel(timestamp, logMessage, normalizedLevel, spanId, traceId)))
```

Comment:
I was able to 100% reproduce logs being dropped in unawaited async scenarios by adding a sleep before this statement. This allowed transaction finalization to occur before AddLogEvent was called on the transaction.